src/Api/PubnubApi/EventEngine/EffectDispatcher.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace PubnubApi.PubnubEventEngine
+{
+	public class EffectDispatcher
+	{
+		public enum DispatcherType
+		{
+			Entry,
+			Exit,
+			Managed
+		}
+		public IPubnubUnitTest PubnubUnitTest { get; set; }
+
+		public Dictionary<EventType, IEffectInvocationHandler> effectActionMap;
+		public EffectDispatcher()
+		{
+			effectActionMap = new Dictionary<EventType, IEffectInvocationHandler>();
+		}
+
+		public async void dispatch(EventType effect, ExtendedState stateContext)
+		{
+			IEffectInvocationHandler? handler;
+			if (effectActionMap.TryGetValue(effect, out handler)) {
+				if (handler != null)
+				{
+					await Task.Factory.StartNew(()=> handler.Start(stateContext, effect)).ConfigureAwait(false);
+				}
+			}
+		}
+
+		public async void dispatch(DispatcherType dispatchType, EventType eventType,List<EffectInvocation> effectInvocations, ExtendedState stateContext)
+		{
+			if (effectInvocations == null || effectInvocations.Count == 0) { return; }
+			foreach (var effect in effectInvocations) {
+				PubnubUnitTest?.EventTypeList?.Add(new KeyValuePair<string, string>("invocation", effect.Name));
+				System.Diagnostics.Debug.WriteLine("Found effect " + effect.Effectype);
+				if (dispatchType == DispatcherType.Exit)
+                {
+					await Task.Factory.StartNew(()=> effect.Handler?.Cancel()).ConfigureAwait(false);
+				}
+				else if (dispatchType == DispatcherType.Entry)
+				{
+					await Task.Factory.StartNew(()=> effect.Handler?.Start(stateContext, eventType)).ConfigureAwait(false);
+				}
+				else if (dispatchType == DispatcherType.Managed)
+				{
+					if (effect is EmitStatus)
+					{
+						((EmitStatus)effect).Announce();
+					}
+					else if (effect is EmitMessages<object>)
+					{
+						((EmitMessages<object>)effect).Announce<string>();
```

Comment:
Since EmitMessage is going to be an effect. We may not want to call specific method inside dispatch method.
EmitMessage effect's handler will take care of it.

To avoid the confusion, We may specify 2 Effect types:  `ManagedEffect`, `UnManagedEffect` // taking example - not sure it's good name or not

So those announce related effects can inherit from UnManagedEffect.

```c#
EmitStatusEffect : UnManagedEffect
EmitMessageEffect: UnManagedEffect
```

and we can define `Run` method on it. as `Start()` is not suitable name for such operations.

Reply:
ok. lets discuss this 1-1.